### PR TITLE
[ESI-13047] Retry db transactions when possible

### PR DESCRIPTION
**Pull request by przemyslaw-dobrowolski-cl**
This commit adds new functions Within and WithinTx that execute a scoped database transaction and retry it whenever a deadlock is detected.

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -43,6 +48,67 @@ type DB interface {
...
+			tx.Close()
+		}
+
+		if err == nil {
+			return nil
+		}
+
+		log.Debug("Within: inner function failed with error: %s", err)
+
+		if strings.Contains(err.Error(), "Error 1213: Deadlock found when trying to get lock; try restarting transaction") {
```
Comment (michalzurawski) [inline db/db.go]: works only for MySQL. i.e. SQLite error message for deadlock is "database is locked"

File: db/db.go [hunk 2]
```diff
@@ -43,6 +48,67 @@ type DB interface {
...
+			// deadlock: retry transaction
+			log.Warning("Within: transaction deadlocked, retrying")
+
+			if retryCounter == DEADLOCK_RETRIES {
```
Comment (michalzurawski) [inline db/db.go]: perhaps
```go
for retryCounter = 0; retryCounter == DEADLOCK_RETRIES; retryCounter++ {
```
would be more readable

File: db/db.go [hunk 3]
```diff
@@ -43,6 +48,67 @@ type DB interface {
...
+				log.Warning("Within: transaction deadlocked, gave up after %d retries", DEADLOCK_RETRIES)
+				return err
+			}
+
+			retryCounter++
+			time.Sleep(time.Duration(DEADLOCK_RETRY_DELAY_MS) * time.Millisecond)
+			continue
```
Comment (michalzurawski) [inline db/db.go]: unnecessary?

File: server/resources/resource_management.go [hunk 4]
```diff
@@ -72,30 +72,25 @@ type ExtensionError struct {
 }
 
 //InTransaction executes function in the db transaction and set it to the context
-func InTransaction(ctx middleware.Context, dataStore db.DB, level transaction.Type, f func() error) error {
+func InTransaction(ctx middleware.Context, dataStore db.DB, level transaction.Type, fn func() error) error {
 	if ctx["transaction"] != nil {
```
Comment (p-kozlowski) [inline server/resources/resource_management.go]: if there's a restart of event handling, make a fresh, deep-copy of the context (extensions are allowed to save some state inside)

File: db/db.go (hunk 1 above)
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: ok

File: db/db.go (hunk 2 above)
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: ok

File: db/db.go (hunk 3 above)
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: ok

File: server/resources/resource_management.go (hunk 4 above)
Comment (przemyslaw-dobrowolski-cl) [inline server/resources/resource_management.go]: ok

**Comment by nati** [comment]
I believe this may cause unexpected behavior of the process.

**Comment by przemyslaw-dobrowolski-cl** [comment]
Added a configuration option to enable and configure automatic retry of transactions.

**Comment by marcin-ptaszynski** [comment]
@przemyslaw-dobrowolski-cl , this requires a rebase.

@nati , could you have a look?

**Comment by morrisson** [comment]
I think this PR is not enough to be matured for merge. There are no new test cases to test this new feature. Actually, the idea of this PR is simply adding retry, but why does this patch produce so many different lines (not only adding indentations)? This really makes it hard to check degradation.

Retrying feature might improve the system availability at some level to users, so I kind of agree retry feature itself. However, how much does it improve? We can have multiple options for DB retries, client side, middleware side (e.g., gohan extension) and app native. With comparing other options, how much does this PR improve? We should clarify it before introducing the complexity to gohan.

**Comment by przemyslaw-dobrowolski-cl** [comment]
rebased

File: db/db.go [hunk 5]
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/cloudwan/gohan/db/transaction"
 	"github.com/cloudwan/gohan/schema"
 	"github.com/cloudwan/gohan/util"
+	"time"
```
Comment (marcin-ptaszynski) [inline db/db.go]: stdlib, should be in different section

File: db/options/options.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package options
+
+import "github.com/cloudwan/gohan/util"
+
+// default transaction retry options
+const (
+	DEFAULT_DEADLOCK_RETRY_TX_INTERVAL = 100
```
Comment (marcin-ptaszynski) [inline db/options/options.go]: is this ms?, should be type time.Duration

Comment (marcin-ptaszynski) [inline db/options/options.go]: I think default should be 0, also randomized option would be nice to have.

File: docs/configuration.md [hunk 7]
```diff
@@ -236,6 +236,24 @@ Note: DO NOT USE * configuraion in production deployment.
...
+
+Gohan supports automatic retry of database transactions that have failed due to
+a database deadlock. Under some circumstances (heavy load, multiple clients)
+it is possible that a deadlock occurs in the database and by default,
+Gohan returns an error via API. To override this behaviour configure
+automatic retry of transactions: set retry count to a value greater than 0
+and optionally configure retry intverval in milliseconds.
+
+```yaml
+  retry_tx:
```
Comment (marcin-ptaszynski) [inline docs/configuration.md]: it's not clear from config, this is only for deadlocks. Description is fine, but config should also be clear.

Comment (marcin-ptaszynski) [inline docs/configuration.md]: it's not shown it's in db section.

File: db/db.go (hunk 5 above)
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: fixed

File: db/options/options.go (hunk 6 above)
Comment (przemyslaw-dobrowolski-cl) [inline db/options/options.go]: fixed

Comment (przemyslaw-dobrowolski-cl) [inline db/options/options.go]: fixed

File: docs/configuration.md (hunk 7 above)
Comment (przemyslaw-dobrowolski-cl) [inline docs/configuration.md]: fixed

Comment (przemyslaw-dobrowolski-cl) [inline docs/configuration.md]: fixed

**Comment by przemyslaw-dobrowolski-cl** [comment]
@morrisson could you take a look at this PR - i've added a test to check if retry really works

File: Makefile [hunk 8]
```diff
@@ -11,7 +11,6 @@ deps:
 
 format:
 	@echo "$(OK_COLOR)==> Formatting$(NO_COLOR)"
-
```
Comment (marcin-ptaszynski) [inline Makefile]: unrelated change leftover

Comment (przemyslaw-dobrowolski-cl) [inline Makefile]: ok, fixed

**Comment by przemyslaw-dobrowolski-cl** [comment]
rebased
@marcin-ptaszynski review comments applied - please review again

File: db/options/options.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+package options
+
+import (
+	"github.com/cloudwan/gohan/util"
+	"time"
+)
+
+// default transaction retry options
+const (
+	DEFAULT_DEADLOCK_RETRY_TX_INTERVAL_MSEC = time.Duration(0) * time.Millisecond
```
Comment (marcin-ptaszynski) [inline db/options/options.go]: If type is Duration, _MSEC suffix is incorrect.

Comment (przemyslaw-dobrowolski-cl) [inline db/options/options.go]: right, fixed

File: db/options/options.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type Options struct {
+	RetryTxCount    int
+	RetryTxInterval time.Duration
+}
+
+func Read(config *util.Config) Options {
+	return Options{
+		RetryTxCount:    config.GetInt("database/deadlock_retry_tx/count", DEFAULT_DEADLOCK_RETRY_TX_COUNT),
```
Comment (przemyslaw-dobrowolski-cl) [inline db/options/options.go]: fixed

File: db/db.go [hunk 11]
```diff
@@ -41,86 +42,161 @@ type DB interface {
...
+		}
+
+		if err == nil {
+			return nil
+		}
+
+		log.Debug("withinTxImpl: inner function failed with error: %s", err)
+
+		if !IsDeadlock(err) {
+			return err
```
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: fixed

File: db/db.go [hunk 12]
```diff
@@ -41,86 +42,161 @@ type DB interface {
...
 	err := db.Connect(dbType, conn, maxOpenConn)
 	if err != nil {
 		return nil, err
 	}
 	return db, nil
 }
 
+//ConnectDB is builder function of DB
+func ConnectDB(dbType, conn string, maxOpenConn int) (DB, error) {
+	return ConnectDBWithOptions(dbType, conn, maxOpenConn, options.Default())
```
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: fixed

File: db/sql/sql.go [hunk 13]
```diff
@@ -311,6 +321,10 @@ func (db *DB) Begin() (tx transaction.Transaction, err error) {
 		closed:         false,
 		isolationLevel: transaction.RepeatableRead,
 	}
+	if os.Getenv("FUZZY_DB_TX") == "true" {
```
Comment (przemyslaw-dobrowolski-cl) [inline db/sql/sql.go]: Fuzzy comes from term 'fuzzy testing' - here we randomly insert errors in DB calls. Do you suggest renaming this var to some other more meaningful name?

**Comment by przemyslaw-dobrowolski-cl** [comment]
@morrisson 
applied comments & rebased

File: db/sql/sql.go (hunk 13 above)
Comment (morrisson) [inline db/sql/sql.go]: Is this used in gohan unit test? Name is ok and functionality is great, but there is no documentation about how to use and when we should use.

File: server/resources/resource_management.go [hunk 14]
```diff
@@ -78,31 +78,40 @@ func measureRequestTime(timeStarted time.Time, requestType string, schemaId stri
...
+//resourceTransactionWithContext executes function in the db transaction and set it to the context
+func resourceTransactionWithContext(ctx middleware.Context, dataStore db.DB, level transaction.Type, fn func() error) error {
 	if ctx["transaction"] != nil {
 		return fmt.Errorf("cannot create nested transaction")
 	}
 
-	aTransaction, err := dataStore.BeginTx(context.Background(), &transaction.TxOptions{IsolationLevel: level})
-	if err != nil {
-		return fmt.Errorf("cannot create transaction: %v", err)
+	org := middleware.Context{}
```
Comment (morrisson) [inline server/resources/resource_management.go]: Too short parameter name so I need time to understand what this is. You should change originalCtx and need proper comment why this is necessary.

File: server/server.go [hunk 15]
```diff
@@ -77,48 +79,49 @@ func (server *Server) mapRoutes() {
...
-	if err != nil {
-		log.Fatal(err)
-	}
-	defer tx.Close()
-	coreSchema, _ := schemaManager.Schema("schema")
-	if coreSchema == nil {
-		log.Fatal("Gohan core schema not found")
-		return
-	}
+	if txErr := db.Within(server.db, func(tx transaction.Transaction) error {
```
Comment (morrisson) [inline server/server.go]: tx is not closed anywhere as far as I understand.

File: server/state_watcher.go [hunk 16]
```diff
@@ -158,75 +158,71 @@ func (watcher *StateWatcher) StateUpdate(event *gohan_sync.Event) error {
...
-		return nil
-	}
-	if newError, ok := event.Data["error"].(string); ok {
-		resourceState.Error = newError
-	}
-	if newState, ok := event.Data["state"].(string); ok {
-		resourceState.State = newState
-	}
+	return db.WithinTx(watcher.db, context.Background(), &transaction.TxOptions{IsolationLevel: transaction.GetIsolationLevel(curSchema, StateUpdateEventName)},
+		func(tx transaction.Transaction) error {
```
Comment (morrisson) [inline server/state_watcher.go]: ditto. is tx closed?

File: db/db.go [hunk 17]
```diff
@@ -41,16 +42,88 @@ type DB interface {
...
+			return true
+		}
+	}
+
+	return false
+}
+
+func withinTxImpl(db DB, beginStrategy func(db DB) (transaction.Transaction, error), fn func(transaction.Transaction) error) error {
+	var tx transaction.Transaction
+	var err error
```
Comment (przemyslaw-dobrowolski-cl) [inline db/db.go]: fixed

File: server/resources/resource_management.go (hunk 14 above)
Comment (przemyslaw-dobrowolski-cl) [inline server/resources/resource_management.go]: fixed

File: server/server.go (hunk 15 above)
Comment (przemyslaw-dobrowolski-cl) [inline server/server.go]: tx is closed inside Within()

**Comment by przemyslaw-dobrowolski-cl** [comment]
@morrisson I fixed the comments, although it is not easy to add a reasonable example, included in gohan, for FUZZY_DB_TX. When this flag is enabled it fails very often since gohan is not ready for deadlocks in many places, not necessarily related to resource management which is in scope of this PR.

**Comment by morrisson** [comment]
approved, but this needs @nati 's approval.

**Comment by przemyslaw-dobrowolski-cl** [comment]
rebased
